org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ToolbarCustomButton.java

```diff
@@ -104,7 +106,70 @@ public void dynamicDisplay(boolean forceValidation) {
 
 		toolbarButton.setVisible(visible);
 	}
-	
+
+	public void pressedLogic() {
+		if (toolbarButton.getParent() == null)
+			return;
+
+		String pressedLogic = mToolbarButton.getPressedLogic();
+
+		if (Util.isEmpty(pressedLogic, true))
+			return;
+
+		ADWindow window = ADWindow.get(windowNo);
+
+		if (window == null)
+			return;
+
+		IADTabpanel adTabpanel = window.getADWindowContent().getADTab().getSelectedTabpanel();
+
+		if (adTabpanel == null || adTabpanel.getRecord_ID() <= 0)
+			return;
+
+		boolean isPressed = validateLogic(pressedLogic, adTabpanel.getTabNo());
+		((ToolBarButton) toolbarButton).setPressed(isPressed);
+	}
+
+	public void readOnlyLogic() {
+		if (toolbarButton.getParent() == null)
+			return;
+
+		String readOnlyLogic = mToolbarButton.getReadOnlyLogic();
+
+		if (Util.isEmpty(readOnlyLogic, true))
+			return;
+
+		ADWindow window = ADWindow.get(windowNo);
+
+		if (window == null)
+			return;
+
+		IADTabpanel adTabpanel = window.getADWindowContent().getADTab().getSelectedTabpanel();
+
+		if (adTabpanel == null || adTabpanel.getRecord_ID() <= 0)
+			return;
+
+		boolean isReadOnly = validateLogic(readOnlyLogic, adTabpanel.getTabNo());
+
+		toolbarButton.setDisabled(isReadOnly);
+	}
+
+	private boolean validateLogic(String logic, int tabNo) {
+		boolean isValid = false;
+
+		if (logic.startsWith("@SQL="))
+		{
+			isValid = Evaluator.parseSQLLogic(logic, Env.getCtx(), windowNo, tabNo, "");
+		}
+		else
+		{
+			logic = Env.parseContext(Env.getCtx(), windowNo, tabNo, logic, false);
```

Comment:
hi @dpansheriya , miss this in my previous patch, this line should be removed as well.

Reply:
@hengsin  this is done.